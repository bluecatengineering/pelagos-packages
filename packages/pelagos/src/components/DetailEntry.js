import PropTypes from 'prop-types';

import useRandomId from '../hooks/useRandomId';

import InfoTooltip from './InfoTooltip';
import './DetailEntry.less';

/** An entry in the details panel. */
const DetailEntry = ({
	id,
	className,
	label,
	infoText,
	infoTextPlacement,
	value,
	direction,
	valueClass,
	valueTitle,
	children,
	...props
}) => {
	id = useRandomId(id);
	return (
		<div className={`DetailEntry${className ? ` ${className}` : ''}`}>
			<div className="DetailEntry__label">
				<h3 className="DetailEntry__title">{label}</h3>
				{infoText && <InfoTooltip text={infoText} placement={infoTextPlacement} />}
			</div>
			<div
				{...props}
				id={id}
				className={`DetailEntry__value DetailEntry--${direction}${valueClass ? ` ${valueClass}` : ''}`}
				title={valueTitle}>
				{children || value}
			</div>
		</div>
	);
};

DetailEntry.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.any,
	/** The label text. */
	label: PropTypes.string.isRequired,
	/** The text of the info tooltip to display. */
	infoText: PropTypes.string,
	/** The placement of the info tooltip relative to the icon. */
	infoTextPlacement: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
	/** The value to be displayed. */
	value: PropTypes.any,
	/** The layout direction for the value. */
	direction: PropTypes.oneOf(['row', 'column']),
	/** The value class name(s). */
	valueClass: PropTypes.string,
	/** The title for the value. */
	valueTitle: PropTypes.string,
	/** The child elements, can be provided instead of the value. */
	children: PropTypes.any,
};

DetailEntry.defaultProps = {
	direction: 'column',
};

export default DetailEntry;
