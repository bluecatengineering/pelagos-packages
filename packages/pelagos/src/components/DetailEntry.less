@import '../../less/fonts';
@import '../../less/spacing';

.DetailEntry {
	margin-top: @sp-16;

	&--row {
		flex-direction: row;
	}

	&__label {
		flex-direction: row;
		align-items: center;
		gap: @sp-08;
	}

	&__title {
		@font-12-400();
		color: var(--text-secondary);
	}

	&__value {
		position: relative;
		margin-top: @sp-04;
	}
}
