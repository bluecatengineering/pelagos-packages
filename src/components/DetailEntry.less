@import '../../less/spacing';

.DetailEntry {
	margin-top: @sp-16;

	&--row {
		flex-direction: row;
	}

	&__label {
		flex-direction: row;
		align-items: center;
		gap: @sp-08;
	}

	&__value {
		position: relative;
		margin-top: @sp-04;
	}
}
