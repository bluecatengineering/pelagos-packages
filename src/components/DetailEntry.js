import PropTypes from 'prop-types';

import Label from './Label';
import './DetailEntry.less';

/** An entry in the details panel. */
const DetailEntry = ({id, className, label, value, direction, valueClass, valueTitle, children, ...props}) => (
	<div className={`DetailEntry${className ? ` ${className}` : ''}`}>
		<Label text={label} htmlFor={id} />
		<div
			{...props}
			id={id}
			className={`DetailEntry__value DetailEntry--${direction}${valueClass ? ` ${valueClass}` : ''}`}
			title={valueTitle}
		>
			{children || value}
		</div>
	</div>
);

DetailEntry.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.any,
	/** The label text. */
	label: PropTypes.string.isRequired,
	/** The value to be displayed. */
	value: PropTypes.any,
	/** The layout direction for the value. */
	direction: PropTypes.oneOf(['row', 'column']),
	/** The value class name(s). */
	valueClass: PropTypes.string,
	/** The title for the value. */
	valueTitle: PropTypes.string,
	/** The child elements, can be provided instead of the value. */
	children: PropTypes.any,
};

DetailEntry.defaultProps = {
	direction: 'column',
};

export default DetailEntry;
