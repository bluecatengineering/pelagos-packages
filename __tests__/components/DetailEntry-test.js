import {shallow} from 'enzyme';

import DetailEntry from '../../src/components/DetailEntry';

jest.unmock('../../src/components/DetailEntry');

describe('DetailEntry', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<DetailEntry
					id="test"
					className="TestClass"
					label="Test Label"
					value="test value"
					valueClass="TestValueClass"
					valueTitle="TestValueTitle"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when children are specified', () => {
			const wrapper = shallow(
				<DetailEntry id="test" className="TestClass" label="Test Label" value="test value" valueClass="TestValueClass">
					Test Children
				</DetailEntry>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when direction is row', () => {
			const wrapper = shallow(
				<DetailEntry
					id="test"
					className="TestClass"
					label="Test Label"
					value="test value"
					direction="row"
					valueClass="TestValueClass"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when className and valueClass are not set', () => {
			const wrapper = shallow(<DetailEntry id="test" label="Test Label" value="test value" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});
});
