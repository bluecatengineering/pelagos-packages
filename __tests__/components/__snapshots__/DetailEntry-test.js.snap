// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailEntry rendering renders expected elements 1`] = `
<div
  className="DetailEntry TestClass"
>
  <div
    className="DetailEntry__label"
  >
    <Label
      htmlFor="test"
      text="Test Label"
    />
  </div>
  <div
    className="DetailEntry__value DetailEntry--column TestValueClass"
    id="test"
    title="TestValueTitle"
  >
    test value
  </div>
</div>
`;

exports[`DetailEntry rendering renders expected elements when children are specified 1`] = `
<div
  className="DetailEntry TestClass"
>
  <div
    className="DetailEntry__label"
  >
    <Label
      htmlFor="test"
      text="Test Label"
    />
  </div>
  <div
    className="DetailEntry__value DetailEntry--column TestValueClass"
    id="test"
  >
    Test Children
  </div>
</div>
`;

exports[`DetailEntry rendering renders expected elements when className and valueClass are not set 1`] = `
<div
  className="DetailEntry"
>
  <div
    className="DetailEntry__label"
  >
    <Label
      htmlFor="test"
      text="Test Label"
    />
  </div>
  <div
    className="DetailEntry__value DetailEntry--column"
    id="test"
  >
    test value
  </div>
</div>
`;

exports[`DetailEntry rendering renders expected elements when direction is row 1`] = `
<div
  className="DetailEntry TestClass"
>
  <div
    className="DetailEntry__label"
  >
    <Label
      htmlFor="test"
      text="Test Label"
    />
  </div>
  <div
    className="DetailEntry__value DetailEntry--row TestValueClass"
    id="test"
  >
    test value
  </div>
</div>
`;

exports[`DetailEntry rendering renders expected elements when infoText is set 1`] = `
<div
  className="DetailEntry"
>
  <div
    className="DetailEntry__label"
  >
    <Label
      htmlFor="test"
      text="Test Label"
    />
    <InfoTooltip
      placement="bottom"
      text="Test info"
    />
  </div>
  <div
    className="DetailEntry__value DetailEntry--column"
    id="test"
  >
    test value
  </div>
</div>
`;
